sql/core/src/main/scala/org/apache/spark/sql/columnar/ColumnAccessor.scala

```diff
@@ -117,9 +135,14 @@ private[sql] object ColumnAccessor {
       case DoubleType => new DoubleColumnAccessor(buf)
       case StringType => new StringColumnAccessor(buf)
       case BinaryType => new BinaryColumnAccessor(buf)
-      case DecimalType.Fixed(precision, scale) if precision < 19 =>
-        new FixedDecimalColumnAccessor(buf, precision, scale)
-      case other => new GenericColumnAccessor(buf, other)
+      case dt: DecimalType if dt.precision < 19 => new CompactDecimalColumnAccessor(buf, dt)
```

Comment:
I think we already have, https://github.com/apache/spark/blob/master/sql/catalyst/src/main/scala/org/apache/spark/sql/types/Decimal.scala#L316.

But it's 18 not 19?